inject balances from json into chainspec

dariolina opened the pull request:
It is required to keep the chain spec file manageable with thousands of balances. The JSON file is NOT final as it only contains farmer rewards and no other allocations (team, fdn, etc.).
Tested on devnet on Oct 9.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/autonomys/subspace/blob/main/CONTRIBUTING.md)

teor2345 committed:
Simplify segment item handling

teor2345 committed:
Produce object mappings incrementally

dariolina committed:
rename gemini to taurus testnet

dariolina committed:
update taurus sudo key

dariolina committed:
disable balance transfers on tauurs

dariolina committed:
inject genesis  balances from json into chainspec

nazar-pc commented on crates/subspace-node/src/genesis_allocations.rs:
```diff
@@ -0,0 +1,35 @@
+use sp_runtime::traits::Zero;
+
+pub fn get_genesis_allocations() -> Vec<(AccountId, Balance)> {
+    let file_content = include_str!("genesis_allocations.json");
```
We typically store such things in a constant at the top of the file, not in local variable, see the chain spec example. Also I'd make this function take contents as an argument with constant that contains JSON being in the chain spec file.

nazar-pc commented on crates/subspace-node/src/genesis_allocations.rs:
```diff
@@ -0,0 +1,35 @@
+    let allocations: Value = serde_json::from_str(file_content)
+        .expect("Failed to parse genesis allocations JSON");
+
+    allocations.as_array()
```
All of these types implement Serde. You can just define the shape and then decode the whole thing without manually iterating over it, for balances you can use `NonZeroU128` to automatically check the invariant.

That is unless you want to have nice error messages here (that should never really happen anyway).

nazar-pc reviewed (commented):
Are you sure we need to include this in the software in JSON? It'll make every node build ~12M larger. Compressing it with zstd will reduce the size to ~4M for example.

teor2345 committed:
Limit CI to 2 hours when a test hangs

teor2345 reviewed (commented):
> Are you sure we need to include this in the software in JSON? It'll make every node build ~12M larger. Compressing it with zstd will reduce the size to ~4M for example.

Can we reduce the file size using arrays rather than key/value maps?
That would look like:
```json
[
  [
    "stBJ4naCdaku9nTXrcQ8m8XQSa4arro9pSdW3eaj6skgW9Nwg",
    "1000000000000000000000"
  ],
  [
    "st9idRv3z62dVwfmjLaNiq1FVb2B2j3giib7ns4Vni1wZ9ePj",
    "1106022470000000000000000"
  ],
  ...
]
```

It also seems like all the amounts end in 15 zeroes, so maybe we could do that multiplication in code?

If this large file isn't final, then compressing it will also add 4 MB to the Git repository every time it changes. If we are going to compress it, there's no need for so much white space, so we could run it through a script that removes all white space before compressing it.

If it's uncompressed, it will add 12 MB to Git, but the diffs will be smaller.

Is there a way to just store the file hash in the node/git, and download the latest file when the node starts? Or distribute the file with the node build?

teor2345 committed:
Deny unsafe object mapping subscriptions

nazar-pc committed:
Switch from Ubuntu 20.04 to 22.04

nazar-pc committed:
Avoid allocation in `CowBytes` if possible

nazar-pc committed:
Simplify `PieceProvider` API

nazar-pc committed:
Introduce constants with recommended limit for cached pieces in a single request

nazar-pc committed:
Add `PieceProvider::get_from_cache()` that supports connected peers and retrieving pieces more efficiently in batches

nazar-pc committed:
Use `Arc<Vec<T>>` in network requests instead of `Vec<T>` for better data sharing

nazar-pc committed:
Delete .gitguardian.yaml

teor2345 commented on crates/subspace-node/src/chain_spec.rs:
```diff
@@ -86,7 +87,7 @@ pub fn gemini_3h_compiled() -> Result<GenericChainSpec, String> {
                 .expect("Wrong root account address");
 
-        let balances = vec![(sudo_account.clone(), 1_000 * SSC)];
+        let balances=get_genesis_allocations(GENESIS_ALLOCATIONS);
```
This should make rustfmt happier
```suggestion
        let balances = get_genesis_allocations(GENESIS_ALLOCATIONS);
```

nazar-pc committed:
Improve docs

Co-authored-by: teor <teor@riseup.net>

teor2345 committed:
Skip empty object mapping notifications

teor2345 committed:
Don't fetch ignored genesis mappings

teor2345 committed:
Skip runtime object mapping calls rather than notifications

teor2345 committed:
Add a --create-object-mappings node CLI option

nazar-pc committed:
Support specifying dialing address for disconnected peer during generic request

nazar-pc committed:
Make requests to peers with explicit addresses succeed

nazar-pc committed:
Use `anyhow` more widely in farmer code

nazar-pc committed:
Delete frame-weight-template.hbs

nazar-pc committed:
Replace `tokio` semaphore with `async-lock`

nazar-pc committed:
Refactor piece retrieval during plotting to be closer in structure to what will be needed later

nazar-pc committed:
Add `PieceGetter::get_pieces` with dummy (for now) implementation

nazar-pc committed:
Take advantage of batched piece downloading in `subspace-farmer-components`

teor2345 committed:
Rewrite README intros to be clearer

teor2345 committed:
Remove deleted res directory in Cargo.toml

teor2345 committed:
Initial subspace-gateway binary which does nothing

nazar-pc committed:
Replace `tokio::sync::mpsc` with `futures::channel::mpsc` in farmer cache

nazar-pc committed:
Add `FarmerCache::get_pieces()` method that can get many pieces concurrently from different caches

nazar-pc committed:
Remove de-duplication of pieces that are being downloaded in `FarmerPieceGetter`

nazar-pc committed:
Inline `get_piece_internal`, no other changes

nazar-pc committed:
Implement `PieceGetter::get_pieces()` in `FarmerPieceGetter`

nazar-pc committed:
Fix disabling of balance transfers

NingLin-P committed:
Add source to the OperatorRewarded event

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Set OnXDMRewards for the consensus test runtime to keep consistency with the production runtime

Signed-off-by: linning <linningde25@gmail.com>

teor2345 committed:
Add basic DSN client implementation

teor2345 committed:
Simplify CLI doc phrasing

teor2345 committed:
Fix node RPC docs

teor2345 committed:
Add initial gateway RPC implementation

teor2345 committed:
Add draft README

nazar-pc committed:
Make sure sector metadata is stored in memory in the correct order during initial plotting

nazar-pc committed:
Minor simplification of `FuturesUnordered` usage

nazar-pc committed:
Unify subscriptions a bit in `NatsClient`

nazar-pc committed:
Introduce `NatsClient::stream_request_responder()` and hide `NatsClient::stream_response()`

nazar-pc committed:
Remove `StreamRequest` and hide stream request/response internals from external users

nazar-pc committed:
Add `PieceCache::read_pieces()` method for more efficient piece cache reads in batches

nazar-pc committed:
Fix resuming future after completion

nazar-pc committed:
Fix docs links to private items

nazar-pc committed:
Avoid division by zero by skipping empty list of peers pieces

nazar-pc committed:
Update `futures` crate

nazar-pc committed:
Fix incorrect `stream::poll_fn` implementation

teor2345 reviewed (commented):
Here's how I checked the balances using `jq`:
```sh
# All balances are 1 or greater
$ cat crates/subspace-node/src/genesis_allocations.json | jq 'map(select(.[1] < 1))
[]
# There are a few large balances
$ ... | jq 'map(select(.[1] > 1000000))'
[
  [
    "st9idRv3z62dVwfmjLaNiq1FVb2B2j3giib7ns4Vni1wZ9ePj",
    1106023
  ],
  [
    "st9wSyY8mWAwhBDnKxjww6onxSHGEf4uAKGWdmGR2fcL87bTY",
    1142142
  ]
]
# And lots of small balances
$ ... | jq 'map(select(.[1] == 1))|length'
36511
# Number of entries
$ ... | jq 'length'
108388
# Number of unique addresses
$ ... | jq 'unique_by(.[0])|length'
108388
```

The only odd thing I found was a suspiciously round number of unique amounts.
```sh
$ ... | jq 'unique_by(.[1])|length'
4000
```

Is it possible that the data generation script hit some kind of limit, and started ignoring different amounts once it had 4000 of them?
(It seems unlikely, but so does having exactly 4000 unique reward amounts.)

Here are some commands to generate an `[amount, count]` histogram:
```sh
# Full histogram
$ cat crates/subspace-node/src/genesis_allocations.json | jq 'group_by(.[1])|map([.[0][1],length])'
# Amounts less than 500 and counts less than 1000
$ ... | jq 'group_by(.[1])|map([.[0][1],length])|map(select([.[0] <= 500,.[1] <= 1000]|all))'
```
And here's part of that histogram, created using a jsontochart.com scatter plot. Amounts are across the bottom, counts are up the side. I limited amounts to 500 and counts to 1000 to make the graph easier to read.
![jsonchart (1)](https://github.com/user-attachments/assets/af783af5-07da-4126-9a49-d0593a248fa6)

Edit: Is this the reward distribution we expect?

teor2345 committed:
Basic gateway RPC and run implementation

teor2345 committed:
Minimise features used by RPC servers

teor2345 committed:
Support IPv6 in CORS rules

nazar-pc committed:
Fix request-response with addresses

teor2345 committed:
Add TODOs for a library part of this crate

teor2345 committed:
Implement DSN ObjectPieceGetter

teor2345 committed:
Improve PieceValidator debugging

teor2345 committed:
Make DSN CLI option names match subspace-node

teor2345 committed:
Log RPC server address

teor2345 committed:
Fix confusing piece getter error name

nazar-pc committed:
Implement `PieceCache::read_pieces()` in everything except `ClusterPieceGetter`

nazar-pc committed:
Small API changes and simplifications in NATS client

nazar-pc committed:
Implement `PieceCache::read_pieces()` for `ClusterPieceGetter`

nazar-pc committed:
Remove piece getter concurrency option

nazar-pc committed:
Download pieces in batches for piece cache sync

teor2345 committed:
Add Node RPC client implementation

teor2345 committed:
Get DSN parameters from node via RPC

teor2345 committed:
Implement piece validation

teor2345 committed:
Fix a typo in farmer piece validation

nazar-pc committed:
Fix typo

nazar-pc committed:
Minor renaming for consistency

nazar-pc committed:
Fix piece retrieval in `PieceProvider`

nazar-pc committed:
Fix cache writes

nazar-pc committed:
Skip peers that already replied to segment headers response successfully

nazar-pc committed:
Small optimization

nazar-pc committed:
Rename CLI options for consistency

nazar-pc committed:
Cleanups in gateway CLI

teor2345 committed:
Remove unnecessary TODOs

teor2345 committed:
Remove unused Error type

teor2345 committed:
Tweak log wording

teor2345 committed:
Rename dsn module to network

teor2345 committed:
Add DSN listen on addresses for debugging

teor2345 committed:
Add a --max-size argument to the gateway

nazar-pc committed:
Avoid replotting the same sector more than once

nazar-pc committed:
Limit per-farm plotting concurrency

nazar-pc committed:
Limit in-flight plotting sectors until the very end

dariolina committed:
prepare taurus chainspec

nazar-pc committed:
Update Rust to the latest nightly

nazar-pc committed:
TODOs cleanup

NingLin-P committed:
Use head domain number to count for the unlocking period

Signed-off-by: linning <linningde25@gmail.com>

nazar-pc committed:
Fix cluster plotting

dariolina committed:
add entropy to taurus chainspec

dariolina committed:
proper formatting

nazar-pc committed:
Update `rustsec/audit-check` action

dariolina committed:
rename gemini-3h to taurus in workflows

nazar-pc committed:
Tiny plotting simplification

nazar-pc committed:
Remove msys64 workaround

nazar-pc committed:
Fix Ubuntu in forks

nazar-pc committed:
Consolidate Docker-related stuff in `docker` directory

teor2345 committed:
Replace panic with error in snap sync

teor2345 committed:
Fix typo in node runner panic message

teor2345 committed:
Exit on panics, rather than unwinding

dariolina committed:
change telemetry endpoint to foundation

nazar-pc committed:
Make farmer reward address optional with `--dev`

nazar-pc committed:
Inherit service span in PoT timekeeper thread

nazar-pc committed:
Minor cleanups

dariolina committed:
add council members into chainspec

nazar-pc committed:
Use hex for human-readable serde encoding and binary otherwise

NingLin-P committed:
Remove the try-runtime feature

It is broken, incomplete, and unused, we can add it back when we have use case.

Signed-off-by: linning <linningde25@gmail.com>

teor2345 committed:
Stay at v3 Cargo.lock file format

NingLin-P committed:
Keep our customized executive pallet up-to-date with the upstream

Signed-off-by: linning <linningde25@gmail.com>

teor2345 committed:
Exit when snap sync needs user intervention

dariolina committed:
rename gemini to taurus in geness storage build

dariolina committed:
proper formatting

nazar-pc committed:
Make `DirectIoFile` API a bit more flexible

nazar-pc committed:
Remove `physical_sector_size` in `DirectIoFile`

nazar-pc committed:
Fix assertion, simplify and unify code between reads and writes

dariolina committed:
permanent bootstrap nodes for taurus

nazar-pc reviewed (commented):
Consider squashing commits. Also what do you think about having one account per line? It shouldn't increase the file size too much, but will make it much easier to handle in text editors and such.

Like this:
```
[
["abc",1000],
["abc",1000],
["abc",1000],
...
]
```
